### ISSUE TITLE
Cannot Use UserInfoRestTemplateCustomizer When Not A Client

### ISSUE THREAD

**Issue by ryanjbaxter** [spring-cloud/spring-cloud-security#50]
I would like to use UserInfoRestTemplateCustomizer in a resource server that is not a client in order to customize the bearer token header sent to the userInfoUri.  Currently I cannot do that because ResourceServerTokenServicesConfiguration it requires that spring.oauth2.client.clientId be set.

**Comment by ryanjbaxter** [comment]
@dsyer I tried the 1.0.1 snapshot and I am still not seeing my customizer get called, here is the bean

```
  @Configuration
  static class OAuth2Config {
    @Bean
    public UserInfoRestTemplateCustomizer userInfoRestTemplate() {
      return new SessionQAUserInfoRestTemplateCustomizer();
    }
  }
```

Does that seem like it should work?

**Comment by dsyer** [comment]
Is is even instantiated (given that the configuration is not visible)?

**Comment by ryanjbaxter** [comment]
OK I got it somewhat working, it was an issue on my end.  However I am running into a NPE, that looks similar to #48 but may not be exactly the same.  I didnt have time to look to deep into it, so it could be another issue on my end, but I don't see any of my code in the stack.

```
2015-03-21 11:22:25.750  INFO 6442 --- [nio-8383-exec-1] o.s.c.s.o.r.UserInfoTokenServices        : Getting user info from: https://graph.facebook.com/v2.2/me/permissions
2015-03-21 11:22:31.371  INFO 6442 --- [nio-8383-exec-1] s.c.a.AnnotationConfigApplicationContext : Refreshing org.springframework.context.annotation.AnnotationConfigApplicationContext@577a2f19: startup date [Sat Mar 21 11:22:31 EDT 2015]; parent: org.springframework.boot.context.embedded.AnnotationConfigEmbeddedWebApplicationContext@4cc963fd
2015-03-21 11:22:31.409  INFO 6442 --- [nio-8383-exec-1] f.a.AutowiredAnnotationBeanPostProcessor : JSR-330 'javax.inject.Inject' annotation found and supported for autowiring
2015-03-21 11:22:31.606  INFO 6442 --- [nio-8383-exec-1] c.netflix.config.ChainedDynamicProperty  : Flipping property: graph.facebook.com.ribbon.ActiveConnectionsLimit to use NEXT property: niws.loadbalancer.availabilityFilteringRule.activeConnectionsLimit = 2147483647
2015-03-21 11:22:31.697  INFO 6442 --- [nio-8383-exec-1] c.netflix.loadbalancer.BaseLoadBalancer  : Client:graph.facebook.com instantiated a LoadBalancer:DynamicServerListLoadBalancer:{NFLoadBalancer:name=graph.facebook.com,current list of Servers=[],Load balancer stats=Zone stats: {},Server stats: []}ServerList:null
2015-03-21 11:22:31.706  INFO 6442 --- [nio-8383-exec-1] c.n.l.DynamicServerListLoadBalancer      : DynamicServerListLoadBalancer for client graph.facebook.com initialized: DynamicServerListLoadBalancer:{NFLoadBalancer:name=graph.facebook.com,current list of Servers=[],Load balancer stats=Zone stats: {},Server stats: []}ServerList:org.springframework.cloud.netflix.ribbon.eureka.DomainExtractingServerList@1b450c26
2015-03-21 11:22:31.777  INFO 6442 --- [nio-8383-exec-1] com.netflix.http4.ConnectionPoolCleaner  : Initializing ConnectionPoolCleaner for NFHttpClient:graph.facebook.com
2015-03-21 11:22:31.972  INFO 6442 --- [nio-8383-exec-1] com.netflix.http4.ConnectionPoolCleaner  : Initializing ConnectionPoolCleaner for NFHttpClient:graph.facebook.com
2015-03-21 11:22:32.212 ERROR 6442 --- [nio-8383-exec-1] o.a.c.c.C.[.[.[/].[dispatcherServlet]    : Servlet.service() for servlet [dispatcherServlet] in context with path [] threw exception

java.lang.NullPointerException: null
	at com.google.common.base.Preconditions.checkNotNull(Preconditions.java:210)
	at com.google.common.cache.LocalCache.get(LocalCache.java:3936)
	at com.google.common.cache.LocalCache.getOrLoad(LocalCache.java:3941)
	at com.google.common.cache.LocalCache$LocalLoadingCache.get(LocalCache.java:4824)
	at com.netflix.loadbalancer.LoadBalancerStats.getServerStats(LoadBalancerStats.java:175)
	at com.netflix.loadbalancer.LoadBalancerStats.getSingleServerStat(LoadBalancerStats.java:362)
	at com.netflix.loadbalancer.LoadBalancerContext.getServerStats(LoadBalancerContext.java:621)
	at org.springframework.cloud.netflix.ribbon.RibbonLoadBalancerClient.execute(RibbonLoadBalancerClient.java:71)
	at org.springframework.cloud.client.loadbalancer.LoadBalancerInterceptor.intercept(LoadBalancerInterceptor.java:46)
	at org.springframework.http.client.InterceptingClientHttpRequest$RequestExecution.execute(InterceptingClientHttpRequest.java:84)
	at org.springframework.cloud.security.oauth2.resource.ResourceServerTokenServicesConfiguration$UserInfoRestTemplateConfiguration$1.intercept(ResourceServerTokenServicesConfiguration.java:119)
	at org.springframework.http.client.InterceptingClientHttpRequest$RequestExecution.execute(InterceptingClientHttpRequest.java:84)
	at org.springframework.http.client.InterceptingClientHttpRequest.executeInternal(InterceptingClientHttpRequest.java:69)
	at org.springframework.http.client.AbstractBufferingClientHttpRequest.executeInternal(AbstractBufferingClientHttpRequest.java:48)
	at org.springframework.http.client.AbstractClientHttpRequest.execute(AbstractClientHttpRequest.java:53)
	at org.springframework.web.client.RestTemplate.doExecute(RestTemplate.java:568)
	at org.springframework.security.oauth2.client.OAuth2RestTemplate.doExecute(OAuth2RestTemplate.java:128)
	at org.springframework.web.client.RestTemplate.execute(RestTemplate.java:529)
	at org.springframework.web.client.RestTemplate.getForEntity(RestTemplate.java:261)
	at org.springframework.cloud.security.oauth2.resource.UserInfoTokenServices.getMap(UserInfoTokenServices.java:105)
	at org.springframework.cloud.security.oauth2.resource.UserInfoTokenServices.loadAuthentication(UserInfoTokenServices.java:58)
	at org.springframework.security.oauth2.provider.authentication.OAuth2AuthenticationManager.authenticate(OAuth2AuthenticationManager.java:83)
	at org.springframework.security.oauth2.provider.authentication.OAuth2AuthenticationProcessingFilter.doFilter(OAuth2AuthenticationProcessingFilter.java:150)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:342)
	at org.springframework.security.web.authentication.logout.LogoutFilter.doFilter(LogoutFilter.java:110)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:342)
	at org.springframework.security.web.header.HeaderWriterFilter.doFilterInternal(HeaderWriterFilter.java:57)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:342)
	at org.springframework.security.web.context.SecurityContextPersistenceFilter.doFilter(SecurityContextPersistenceFilter.java:87)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:342)
	at org.springframework.security.web.context.request.async.WebAsyncManagerIntegrationFilter.doFilterInternal(WebAsyncManagerIntegrationFilter.java:50)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:342)
	at org.springframework.security.web.FilterChainProxy.doFilterInternal(FilterChainProxy.java:192)
	at org.springframework.security.web.FilterChainProxy.doFilter(FilterChainProxy.java:160)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:239)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:206)
	at org.springframework.web.filter.CharacterEncodingFilter.doFilterInternal(CharacterEncodingFilter.java:88)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:239)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:206)
	at org.springframework.boot.actuate.autoconfigure.MetricFilterAutoConfiguration$MetricsFilter.doFilterInternal(MetricFilterAutoConfiguration.java:90)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:239)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:206)
	at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:219)
	at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:106)
	at org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:501)
	at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:142)
	at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:79)
	at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:88)
	at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:537)
	at org.apache.coyote.http11.AbstractHttp11Processor.process(AbstractHttp11Processor.java:1085)
	at org.apache.coyote.AbstractProtocol$AbstractConnectionHandler.process(AbstractProtocol.java:658)
	at org.apache.coyote.http11.Http11NioProtocol$Http11ConnectionHandler.process(Http11NioProtocol.java:222)
	at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1556)
	at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.run(NioEndpoint.java:1513)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
	at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61)
	at java.lang.Thread.run(Thread.java:744)
```

**Comment by ryanjbaxter** [comment]
@dsyer so I looked at this exception a little more and it looks like we are trying to use Ribbon to load balance for a service called "graph.facebook.com".  Obviously this service does not exist, hence the NPE.  Should we be trying to load balance requests to the OAuth provider?  It looks like `LoadBalancerInterceptor` is intercepting all HTTP requests and trying to load balance them?

**Comment by dsyer** [comment]
We had some issues with ribbon and OAuth so we disabled it (before 1.0.0 I think). What version are you using?

**Comment by ryanjbaxter** [comment]
I am using 1.0.1 snapshot

**Comment by ryanjbaxter** [comment]
@dsyer any thoughts?  Was this code enabled again in the 1.0.1 snapshots?  Should I create another issue?

**Comment by dsyer** [comment]
I think there is an issue (#51).

**Comment by spencergibb** [comment]
@ryanjbaxter can you see at what point `InterceptingHttpAccessor.setInterceptors` is called that adds the  `LoadBalancerInterceptor`?  We don't add it except for the `RestTemplate` we create in s-c-commons and we don't add it to `OAuth2RestTemplate`.

**Comment by ryanjbaxter** [comment]
@spencergibb if I understand what you are asking, I believe I see it being added from here https://github.com/spring-cloud/spring-cloud-security/blob/8e2c8f73a1d9d1dbd2d7555483fdc3d869dd7006/src/main/java/org/springframework/cloud/security/oauth2/client/OAuth2LoadBalancerClientAutoConfiguration.java#L52

**Comment by spencergibb** [comment]
@ryanjbaxter  @dsyer added it back here: 6da29f42810993e7fa5a431d4b61c28edc4f4a81  Looking at his comment, it looks like you will need to create your own `OAuth2RestTemplate` that doesn't use ribbon.  I'll try it on https://github.com/spencergibb/myfeed since I'd have a similar problem I think.

**Comment by ryanjbaxter** [comment]
@spencergibb @dsyer in what cases does it make sense to use the Ribbon load balancer for OAuth requests?  The only thing I can think of is if the OAuth provider was a service registered with Eureka and not something external (ie GitHub, Facebook, Google, etc)

**Comment by spencergibb** [comment]
@ryanjbaxter I think that is the case.  Hoping to hear from @dsyer as he put it back in.

**Comment by dsyer** [comment]
That's basically the problem. If the authentication comes from an oauth resource on the provider then you definitely don't want ribbon for that bit. But you probably do want it if you have an Oauth resource as a backend service. We need to review this again before 1.0.1.

**Comment by ryanjbaxter** [comment]
So is the only way to fix this in the 1.0.1 snapshots to use my own OAuth2RestTemplate like @spencergibb suggests, or is there an easier way?

**Comment by spencergibb** [comment]
Maybe a config property?  Seems like you'll know if you'll want ribbon or not upfront.

**Comment by spencergibb** [comment]
@ryanjbaxter #58 is an attempt for fix some of your issues

**Comment by ryanjbaxter** [comment]
@spencergibb great!  Looking forward to it getting merged so I move off of 1.0.0 RC3 :)
